core/drivers/d06_lpc_uart.c

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, HiSilicon Limited
+ */
+
+#include <assert.h>
+#include <drivers/d06_lpc_uart.h>
+#include <io.h>
+#include <keep.h>
+#include <mm/core_mmu.h>
+#include <util.h>
+
+static paddr_t lpc_base = LPC_BASE;
+
+static paddr_t chip_to_base(struct serial_chip *chip)
+{
+	struct d06_lpc_uart_data *pd =
+		container_of(chip, struct d06_lpc_uart_data, chip);
+
+	return pd->base.pa;
+}
+
+static void update_lpc_base(void)
+{
+	if (cpu_mmu_enabled())
+		lpc_base = (paddr_t)phys_to_virt_io((paddr_t)LPC_BASE, 0x1000);
+	else
+		lpc_base = LPC_BASE;
+}
+
+static vaddr_t lpc_adjust(paddr_t addr)
+{
+	paddr_t off;
+
+	off = addr - LPC_BASE;
+
+	return (vaddr_t)(off + lpc_base);
+}
+
+static void lpc_byte_read(paddr_t addr, unsigned char *data)
+{
+	LPC_CMD_STRUCT lpc_cmd = {.u32 = 0};
+	unsigned int cnt = 0;
+	unsigned int status;
+
+	lpc_cmd.bits.lpc_wr = 0;
+	lpc_cmd.bits.lpc_type = LPC_TYPE_IO;
+	lpc_cmd.bits.same_addr = LPC_ADDR_MODE_SINGLE;
+	io_write32(lpc_adjust(LPC_CMD_REG), lpc_cmd.u32);
+
+	io_write32(lpc_adjust(LPC_OP_LEN_REG), 1);
+	io_write32(lpc_adjust(LPC_ADDR_REG), addr);
+
+	io_write32(lpc_adjust(LPC_START_REG), 1);
+
+	status = io_read32(lpc_adjust(LPC_IRQ_ST_REG));
+	while (!(status & 0x2)) {
+		if (cnt > 100000)
+			return;
+		cnt++;
+		status = io_read32(lpc_adjust(LPC_IRQ_ST_REG));
+	}
+
+	io_write32(lpc_adjust(LPC_IRQ_ST_REG), 0x2);
+
+	if (io_read32(lpc_adjust(LPC_OP_STATUS_REG)) & 0x2)
+		*data = io_read32(lpc_adjust(LPC_RDATA_REG));
+}
+
+static void lpc_byte_write(paddr_t addr, unsigned char data)
+{
+	LPC_CMD_STRUCT lpc_cmd = {.u32 = 0};
+	unsigned int cnt = 0;
+	unsigned int status;
+
+	lpc_cmd.bits.lpc_wr = 1;
+	lpc_cmd.bits.lpc_type = LPC_TYPE_IO;
+	lpc_cmd.bits.same_addr = LPC_ADDR_MODE_SINGLE;
+	io_write32(lpc_adjust(LPC_CMD_REG), lpc_cmd.u32);
+
+	io_write32(lpc_adjust(LPC_OP_LEN_REG), 1);
+	io_write32(lpc_adjust(LPC_WDATA_REG), data);
+
+	io_write32(lpc_adjust(LPC_ADDR_REG), addr);
+
+	io_write32(lpc_adjust(LPC_START_REG), 1);
+
+	status = io_read32(lpc_adjust(LPC_IRQ_ST_REG));
+	while (!(status & 0x2)) {
+		if (cnt > 100000)
+			return;
+		cnt++;
+		status = io_read32(lpc_adjust(LPC_IRQ_ST_REG));
+	}
+
+	io_write32(lpc_adjust(LPC_IRQ_ST_REG), 0x2);
+}
+
+static void d06_lpc_uart_core_putc(paddr_t base, int ch)
+{
+	unsigned int loop = 0;
+	unsigned char var;
+
+	while (loop < UART_SEND_DELAY) {
```

Comment:
_DELAY is a weird choice since it is not related to time, how about _LOOP_MAX instead?

Reply:
fixed. thx